src/guides/libraries/c-interop.md

```diff
@@ -253,6 +253,65 @@ To reduce this burden,
+
+* It builds (if needed) the native code or
+  obtains the binaries using a package's `build.dart` script.
+* It bundles the native [`Asset`][] reported by the `build.dart` script.
+* It makes the native assets available at runtime through
+  declarative `@Native<>() extern` functions using the [`assetId`][].
+
+The `flutter run` / `flutter build` and `dart` / `dart build` tools
+will now build and bundle native code when
+[opted in](#experiment-opt-in) to the native experiment.
+
+### Walkthrough of `native_add_library`
+
+The [`native_add_library`][] example has the minimum necessary code for
+building and bundling C code in a Dart package.
+
+The example has the following files:
+
+| **Source file**                                                                                                                                                          | **Description**                                                                                                                                                                |
+|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| [`src/native_add_library.c`](https://github.com/dart-lang/native/blob/main/pkgs/native_assets_cli/example/native_add_library/src/native_add_library.c)                   | The C file containing the code for `add`.                                                                                                                                      |
+| [`lib/native_add_library.dart`](https://github.com/dart-lang/native/blob/main/pkgs/native_assets_cli/example/native_add_library/lib/native_add_library.dart)             | The Dart file that invokes the C function `add` in asset `package:native_add_library/native_add_library.dart` through FFI. (Note that _asset id_ defaults to the library uri.) |
+| [`test/native_add_library_test.dart`](https://github.com/dart-lang/native/blob/main/pkgs/native_assets_cli/example/native_add_library/test/native_add_library_test.dart) | A Dart test using the native code.                                                                                                                                             |
+| [`build.dart`](https://github.com/dart-lang/native/blob/main/pkgs/native_assets_cli/example/native_add_library/build.dart)                                               | A script for compiling `src/native_add_library.c` and declaring the compiled asset with  id `package:native_add_library/native_add_library.dart`.                              |
+{:.table .table-striped }
+
+When a Dart or Flutter project depends on `package:native_add_library`,
+the `build.dart` script will automatically be
+invoked on `run`, `build`, and `test` commands.
```

Comment:
Does this apply to these three subcommands for both `dart` and `flutter`? What about `dart <>` without `run`?

Also, does the experiment not work with `dart compile`?

Reply:
`dart compile` creates a single file, and native assets are files. Therefore `dart build` creates a directory rather than a file. `dart compile` gives an error if there are native assets now. (In some future we might be able to statically link the native code all in to one executable, but that's not yet possible.)

`dart path/to/file` works after having done `dart run`:

```
dacoharkes-macbookpro2:native_add_app dacoharkes$ pwd
/Users/dacoharkes/src/dart-lang/native/pkgs/native_assets_cli/example/native_add_app
dacoharkes-macbookpro2:native_add_app dacoharkes$ dart --enable-experiment=native-assets bin/native_add_app.dart 
Invoking a native function to calculate 1 + 2.
Unhandled exception:
Invalid argument(s): Couldn't resolve native function 'add' in 'package:native_add_library/native_add_library.dart' : dlsym(RTLD_DEFAULT, add): symbol not found.

#0      FfiTrampoline__ffi_resolver (dart:ffi)
#1      _add$Method$FfiNative$Ptr (package:native_add_library/native_add_library.dart:12:14)
#2      _add$Method$FfiNative$Ptr (package:native_add_library/native_add_library.dart)
#3      add (package:native_add_library/native_add_library.dart)
#4      main (file:///Users/dacoharkes/src/dart-lang/native/pkgs/native_assets_cli/example/native_add_app/bin/native_add_app.dart:9:18)
#5      _delayEntrypointInvocation.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:297:19)
#6      _RawReceivePort._handleMessage (dart:isolate-patch/isolate_patch.dart:184:12)
dacoharkes-macbookpro2:native_add_app dacoharkes$ dart --enable-experiment=native-assets run
Building package executable... 
Built native_add_app:native_add_app.
Invoking a native function to calculate 1 + 2.
Invocation success: 1 + 2 = 3.
dacoharkes-macbookpro2:native_add_app dacoharkes$ dart --enable-experiment=native-assets bin/native_add_app.dart 
Invoking a native function to calculate 1 + 2.
Invocation success: 1 + 2 = 3.
```

It then uses the cached native assets builds.

Reply:
As long as `dart compile` has an appropriate error, which it seems to as you mentioned, I'm happy. Thanks for the clarification!

https://github.com/dart-lang/sdk/blob/3c456ccbd225a1655c9fdf2eb7c9e452fec2b632/pkg/dartdev/lib/src/commands/compile.dart#L339